File: composeApp/src/commonMain/kotlin/com/jankinwu/fntv/client/data/network/impl/FnApiHelper.kt
package com.jankinwu.fntv.client.data.network.impl

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.jankinwu.fntv.client.BuildConfig
import com.jankinwu.fntv.client.data.store.AppSettingsStore
import korlibs.crypto.MD5
import korlibs.crypto.SHA256
import kotlin.random.Random
import kotlin.time.Clock
import kotlin.time.ExperimentalTime

object FnApiHelper {
    private const val API_KEY = "NDzZTVxnRKP8Z0jXg1VAMonaG8akvh"
    private const val API_SECRET = "16CCEB3D-AB42-077D-36A1-F355324E4237"

    val mapper = jacksonObjectMapper().apply {
        // 禁止格式化输出
        disable(SerializationFeature.INDENT_OUTPUT)
        // 忽略未知字段
        disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        // 不序列化null值
        disable(SerializationFeature.WRITE_NULL_MAP_VALUES)
//            setSerializationInclusion(JsonInclude.Include.NON_NULL)
    }

    @OptIn(ExperimentalTime::class)
    fun genAuthxForOfficial(
        url: String,
        parameters: Map<String, Any?>? = null,
        data: Any? = null,
    ): String {
        return genAuthxInternal(
            url = url,
            parameters = parameters,
            data = data,
            apiSecret = API_SECRET
        )
    }

    @OptIn(ExperimentalTime::class)
    fun genAuthxForFlyNarwhal(
        url: String,
        parameters: Map<String, Any?>? = null,
        data: Any? = null,
    ): String {
        val configuredSecret = BuildConfig.FLY_NARWHAL_API_SECRET.trim()
        return genAuthxInternal(
            url = url,
            parameters = parameters,
            data = data,
            apiSecret = configuredSecret.ifBlank { API_SECRET }
        )
    }

    @OptIn(ExperimentalTime::class)
    fun genAuthx(
        url: String,
        parameters: Map<String, Any?>? = null,
        data: Any? = null,
    ): String {
        return genAuthxForFlyNarwhal(url = url, parameters = parameters, data = data)
    }

    @OptIn(ExperimentalTime::class)
    private fun genAuthxInternal(
        url: String,
        parameters: Map<String, Any?>?,
        data: Any?,
        apiSecret: String
    ): String {
        val nonce = generateRandomDigits()
        val timestamp = Clock.System.now().toEpochMilliseconds().toString()
        val dataJsonMd5 = buildDataJsonMd5(parameters, data)

        val signList = mutableListOf(
            API_KEY,
            url,
            nonce,
            timestamp,
            dataJsonMd5,
            apiSecret
        )

        val signStr = signList.joinToString("_")
        val sign = getMd5(signStr)
        return "nonce=$nonce&timestamp=$timestamp&sign=${sign}"
    }

    fun genSignxForFlyNarwhal(
        url: String,
        authx: String,
        parameters: Map<String, Any?>? = null,
        data: Any? = null,
        publicKeyBase64: String = AppSettingsStore.authCode
    ): String {
        val authxMap = parseAuthxHeader(authx)
        val nonce = authxMap["nonce"].orEmpty()
        val timestamp = authxMap["timestamp"].orEmpty()
        val sign = authxMap["sign"].orEmpty()
        val dataJsonMd5 = buildDataJsonMd5(parameters, data)
        val signxStr = listOf(timestamp, nonce, sign, dataJsonMd5, url, publicKeyBase64).joinToString("_")
        return sha256Hex(signxStr)
    }

    private fun generateRandomDigits(start: Int = 100000, end: Int = 1000000): String {
        return Random.nextInt(start, end).toString()
    }

    private fun buildDataJsonMd5(parameters: Map<String, Any?>?, data: Any?): String {
        return when {
            data != null -> {
                val dataJson = mapper.writeValueAsString(data)
                getMd5(dataJson)
            }

            parameters != null -> {
                val sortedParams = parameters.filterValues { it != null }
                    .toSortedMap()
                    .map { "${it.key}=${it.value}" }
                    .joinToString("&")
                getMd5(sortedParams)
            }

            else -> getMd5("")
        }
    }

    private fun parseAuthxHeader(authx: String): Map<String, String> {
        return authx.split("&")
            .mapNotNull { part ->
                val kv = part.split("=", limit = 2)
                if (kv.size == 2) kv[0] to kv[1] else null
            }
            .toMap()
    }

    private fun getMd5(input: String): String {
        return MD5.digest(input.toByteArray(Charsets.UTF_8)).hex
    }

    private fun sha256Hex(input: String): String {
        return SHA256.digest(input.toByteArray(Charsets.UTF_8)).hex
    }
}

File: lib/utils/fn_api_helper.dart
import 'dart:convert';
import 'dart:math';
import 'package:crypto/crypto.dart';
import 'package:cryptography/cryptography.dart';

class FnApiHelper {
  static const String _apiKey = "NDzZTVxnRKP8Z0jXg1VAMonaG8akvh";
  static const String _apiSecret = "16CCEB3D-AB42-077D-36A1-F355324E4237";

  static String genAuthxForOfficial(
    String url, {
    Map<String, dynamic>? parameters,
    dynamic data,
  }) {
    return _genAuthxInternal(
      url: url,
      parameters: parameters,
      data: data,
      apiSecret: _apiSecret,
    );
  }

  static String genAuthxForFlyNarwhal(
    String url, {
    Map<String, dynamic>? parameters,
    dynamic data,
    String? apiSecret,
  }) {
    return _genAuthxInternal(
      url: url,
      parameters: parameters,
      data: data,
      apiSecret: (apiSecret ?? _apiSecret).trim().isEmpty ? _apiSecret : apiSecret!,
    );
  }

  static String genSignxForFlyNarwhal({
    required String url,
    required String authx,
    Map<String, dynamic>? parameters,
    dynamic data,
    required String publicKeyBase64,
  }) {
    final authxMap = _parseAuthxHeader(authx);
    final nonce = authxMap['nonce'] ?? '';
    final timestamp = authxMap['timestamp'] ?? '';
    final sign = authxMap['sign'] ?? '';
    final dataJsonMd5 = _buildDataJsonMd5(parameters, data);
    final signxStr = [
      timestamp,
      nonce,
      sign,
      dataJsonMd5,
      url,
      publicKeyBase64,
    ].join("_");
    return _sha256Hex(signxStr);
  }

  static Future<String> clientKeyxBase64Url() {
    return FlyNarwhalKeyExchange.clientKeyxBase64Url();
  }

  static String _genAuthxInternal({
    required String url,
    Map<String, dynamic>? parameters,
    dynamic data,
    required String apiSecret,
  }) {
    final nonce = _generateRandomDigits();
    final timestamp = DateTime.now().millisecondsSinceEpoch.toString();
    final dataJsonMd5 = _buildDataJsonMd5(parameters, data);

    final signList = [
      _apiKey,
      url,
      nonce,
      timestamp,
      dataJsonMd5,
      apiSecret,
    ];

    final signStr = signList.join("_");
    final sign = _getMd5(signStr);
    return "nonce=$nonce&timestamp=$timestamp&sign=$sign";
  }

  static String _generateRandomDigits({int start = 100000, int end = 1000000}) {
    final random = Random();
    return (start + random.nextInt(end - start)).toString();
  }

  static String _buildDataJsonMd5(Map<String, dynamic>? parameters, dynamic data) {
    if (data != null) {
      final dataJson = jsonEncode(data);
      return _getMd5(dataJson);
    } else if (parameters != null) {
      final sortedKeys = parameters.keys.toList()..sort();
      final sortedParams = sortedKeys
          .where((key) => parameters[key] != null)
          .map((key) => "$key=${parameters[key]}")
          .join("&");
      return _getMd5(sortedParams);
    } else {
      return _getMd5("");
    }
  }

  static Map<String, String> _parseAuthxHeader(String authx) {
    final parts = authx.split("&");
    final map = <String, String>{};
    for (final part in parts) {
      final kv = part.split("=");
      if (kv.length == 2) {
        map[kv[0]] = kv[1];
      }
    }
    return map;
  }

  static String _getMd5(String input) {
    return md5.convert(utf8.encode(input)).toString();
  }

  static String _sha256Hex(String input) {
    return sha256.convert(utf8.encode(input)).toString();
  }
}

class FlyNarwhalKeyExchange {
  static final X25519 _algorithm = X25519();
  static SimpleKeyPair? _keyPair;
  static SimplePublicKey? _publicKey;

  static Future<String> clientKeyxBase64Url() async {
    _keyPair ??= await _algorithm.newKeyPair();
    _publicKey ??= await _keyPair!.extractPublicKey();
    final raw = _publicKey!.bytes;
    final encoded = base64UrlEncode(raw);
    return encoded.replaceAll('=', '');
  }
}
